### PR TITLE
feat: only re-highlight changed ranges

### PR DESCRIPTION
**Pull request by xcb-xwii**
Requires https://github.com/euclidianAce/ltreesitter/pull/12.

### CODE REVIEW

File: init.lua [hunk 1]
```diff
@@ -123,6 +123,18 @@ local function accumulateLen(tbl)
 	return len
 end
 
+local function incrementalHighlight(doc)
+	local old = doc.ts.tree
+	doc.ts.tree = doc.ts.parser:parse_with(tsInput(doc.lines), doc.ts.tree)
+
+	for _, p in ipairs(old:get_changed_ranges(doc.ts.tree)) do
+		for i=p.start_point.row+1,p.end_point.row+1 do
```
Comment (TorchedSammy) [inline init.lua:131]: add spacing

File: init.lua [hunk 2]
```diff
@@ -193,25 +201,31 @@ local oldTokenize = Highlight.tokenize_line
...
 	local tokens = res.tokens
 	local currentLine = self.doc.lines[idx]
-	local lastNode, lastStartPoint, lastEndPoint
-	for n, nName in self.doc.ts.query:capture(self.doc.ts.tree:root()) do
+	local lastNode, lastName, lastStartPoint, lastEndPoint
+	for
+		n, nName in
+		self.doc.ts.query:capture(self.doc.ts.tree:root(),
+			{ row = i, column = 0 }, { row = i, column = #currentLine - 1 })
+	do
```
Comment (TorchedSammy) [inline init.lua:224]: could you only line break in the table

File: init.lua [hunk 3]
```diff
@@ -224,16 +238,25 @@ function Highlight:tokenize_line(idx, state)
...
 		if startPoint.row == endPoint.row then
-			local append_idx =
-				(lastNode and lastStartPoint.column == startPoint.column and lastEndPoint.column == endPoint.column) -- if the nodes overlap
-					and (#tokens - 1) -- replace the old one
-					or (#tokens + 1)  -- create a new token
-			tokens[append_idx] = nName
-			tokens[append_idx+1] = currentLine:sub(startPoint.column + 1, endPoint.column)
-
+			if lastNode and isBefore(startPoint, lastEndPoint) then
+				if (lastName ~= "error") then
```
Comment (TorchedSammy) [inline init.lua:242]: braces here not needed

**Review by TorchedSammy — CHANGES_REQUESTED** [review]
a little bit of style changes but otherwise looks good

**Review by TorchedSammy — CHANGES_REQUESTED** [review]
actually it seems like this introduces odd artifacts:

**Comment by xcb-xwii** [comment]
> actually it seems like this introduces odd artifacts:  
> 
> 

Have a fix, will push later

**Comment by xcb-xwii** [comment]
Resolved the issue with the artefacts, and rebased onto master.

**Review by TorchedSammy — APPROVED** [review]

**Pull request merged**